[WebDriver][BiDi] Assertion in RefCountedAndCanMakeWeakPtr<WebSocketServer> destructor

lauromoura opened the pull request:
#### 71b34966f4fbbb0e2c7bbf5132c4645a0ef67e90
<pre>
[WebDriver][BiDi] Assertion in RefCountedAndCanMakeWeakPtr&lt;WebSocketServer&gt; destructor
<a href="https://bugs.webkit.org/show_bug.cgi?id=291925">https://bugs.webkit.org/show_bug.cgi?id=291925</a>

Reviewed by BJ Burg and Carlos Garcia Campos.

Make WebDriverService hold the WebSocketServer in a Ref, as the server
implements RefCountedAndCanMakeWeakPtr.

This commit also removes the member of WebSocketServer that used to hold
the WebDriverService, as it was just used in
`WebSocketServer::session(...)`. This method was changed to return the
session id associated to that connection, so we don&apos;t have to pass the
current session anymore.

Combined changes:
* Source/WebDriver/WebDriverService.cpp:
(WebDriver::WebDriverService::WebDriverService): Update server
constructor call.
(WebDriver::WebDriverService::run): Fix pointer usage.
(WebDriver::WebDriverService::acceptHandshake): Ditto and update method
call.
(WebDriver::WebDriverService::handleMessage): Ditto and cleanup.
(WebDriver::WebDriverService::newSession): Fix pointer usage.
(WebDriver::WebDriverService::createSession): Ditto.
(WebDriver::WebDriverService::deleteSession): Ditto.
(WebDriver::WebDriverService::clientDisconnected): Ditto.
(WebDriver::WebDriverService::onBrowserTerminated): Ditto.
* Source/WebDriver/WebDriverService.h: Fix m_bidiServer type.
* Source/WebDriver/WebSocketServer.cpp:
(WebDriver::WebSocketServer::WebSocketServer): Update constructor call.
(WebDriver::WebSocketServer::sessionID const): Added.
(WebDriver::WebSocketServer::session): Deleted.
* Source/WebDriver/WebSocketServer.h: Remove m_bidiService and make
  constructor private.
(WebDriver::WebSocketServer::create): Added.
* Source/WebDriver/soup/WebSocketServerSoup.cpp:
(WebDriver::WebSocketServer::disconnect): Checks and disconnect the
signal handlers.
(WebDriver::WebSocketServer::disconnectSession): Ditto.

Canonical link: <a href="https://commits.webkit.org/294135@main">https://commits.webkit.org/294135@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c68bd01e0e0e62ad30036ee4f433463dd031d4e1

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/100927 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/20589 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/10892 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/106073 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/51525 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/20897 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/29083 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/76860 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/33888 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/103934 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/16072 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/91159 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/57207 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/15887 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/9181 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/50901 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/85796 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/9242 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/108429 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/28054 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/20629 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/85824 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/28417 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/87359 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/85365 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21729 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/30079 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/7804 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/22084 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/27985 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/33253 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/27797 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/31117 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/29355 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on Source/WebDriver/WebSocketServer.cpp:
```diff
@@ -71,7 +70,7 @@ void WebSocketServer::removeConnection(const WebSocketMessageHandler::Connection
 }
 
-RefPtr<Session> WebSocketServer::session(const WebSocketMessageHandler::Connection& connection)
+RefPtr<Session> WebSocketServer::session(const WebSocketMessageHandler::Connection& connection, RefPtr<Session> existingSession)
```
This method is super confusing. it receives a session and it always returns that session or nullptr. I guess what we want here is checking the service session is handled by the server? If that's the case, I think this could be renamed as checkSession() or something like that and return a bool. Another option would be to return the session id, renaming it was sessionID. then the caller can check if it matches whatever session.

lauromoura replied:
Thanks for the suggestion. I'm going with the approach of returning the `SessionID`.

carlosgcampos commented on Source/WebDriver/soup/WebSocketServerSoup.cpp:
```diff
@@ -204,10 +214,11 @@ void WebSocketServer::disconnectSession(const String& sessionId)
     RELEASE_LOG(WebDriverBiDi, "WebSockets support not implemented yet with libsoup2");
 #else
     auto connection = this->connection(sessionId);
-    if (!connection)
+    if (!connection || !connection->get())
```
GRefPtr evaluates to false if it contains a nullptr value, so don't need to change anything here.

lauromoura replied:
Yes, but in this case, `connection` is an `std::optional`, wrapping the `GRefPtr`. So, we might have the - unexpected - case of having a non-empty `std::optional` but pointing to a null `GRefPtr`.

The reason for the optional is that while it could have been a plain `GRefPtr`, I wanted to leave the `WebSocketServer` API open for other ports that might have non-nullable `Connection` types.

FTR, in https://github.com/WebKit/WebKit/pull/42968 we're discussing hiding the `Connection` type at all from `WebSocketServer` public API.

carlosgcampos replied:
Ah, I didn't know it was optional, auto sometimes makes reviews more difficult

carlosgcampos commented on Source/WebDriver/WebDriverService.cpp:
```diff
@@ -452,16 +452,14 @@ void WebDriverService::handleMessage(WebSocketMessageHandler::Message&& message,
+    else {
+        // FIXME Remove once we support checking static vs non-static methods https://bugs.webkit.org/show_bug.cgi?id=281721
+        completionHandler(WebSocketMessageHandler::Message::fail(CommandResult::ErrorCode::InvalidSessionID, connection));
+        return;
```
Why do we need the local variable if we are always returning when not using m_session. Can't we just return and use m_session otherwise?

```
if (m_bidiServer->sessionID(connection) != m_session->id()) {
    // FIXME Remove once we support checking static vs non-static methods https://bugs.webkit.org/show_bug.cgi?id=281721
    completionHandler(WebSocketMessageHandler::Message::fail(CommandResult::ErrorCode::InvalidSessionID, connection));
    return;
}
```

carlosgcampos commented on Source/WebDriver/WebSocketServer.cpp:
```diff
@@ -71,9 +70,9 @@ void WebSocketServer::removeConnection(const WebSocketMessageHandler::Connection
+String WebSocketServer::sessionID(const WebSocketMessageHandler::Connection& connection) const
 {
-    String sessionId;
+    String sessionId = nullString();
```
I think this this is needed, it's already null no?

carlosgcampos commented on Source/WebDriver/WebSocketServer.cpp:
```diff
@@ -82,14 +81,7 @@ RefPtr<Session> WebSocketServer::session(const WebSocketMessageHandler::Connecti
-        return { };
-
-    return existingSession;
+    return sessionId;
```
Instead of breaking in the loop we can just return pair.value and then we don't need the local variable, here we can just return { }

lauromoura committed:
[WebDriver][BiDi] Assertion in RefCountedAndCanMakeWeakPtr<WebSocketServer> destructor
https://bugs.webkit.org/show_bug.cgi?id=291925

Reviewed by BJ Burg and Carlos Garcia Campos.

Make WebDriverService hold the WebSocketServer in a Ref, as the server
implements RefCountedAndCanMakeWeakPtr.

This commit also removes the member of WebSocketServer that used to hold
the WebDriverService, as it was just used in
`WebSocketServer::session(...)`. This method was changed to return the
session id associated to that connection, so we don't have to pass the
current session anymore.

Combined changes:
* Source/WebDriver/WebDriverService.cpp:
(WebDriver::WebDriverService::WebDriverService): Update server
constructor call.
(WebDriver::WebDriverService::run): Fix pointer usage.
(WebDriver::WebDriverService::acceptHandshake): Ditto and update method
call.
(WebDriver::WebDriverService::handleMessage): Ditto and cleanup.
(WebDriver::WebDriverService::newSession): Fix pointer usage.
(WebDriver::WebDriverService::createSession): Ditto.
(WebDriver::WebDriverService::deleteSession): Ditto.
(WebDriver::WebDriverService::clientDisconnected): Ditto.
(WebDriver::WebDriverService::onBrowserTerminated): Ditto.
* Source/WebDriver/WebDriverService.h: Fix m_bidiServer type.
* Source/WebDriver/WebSocketServer.cpp:
(WebDriver::WebSocketServer::WebSocketServer): Update constructor call.
(WebDriver::WebSocketServer::sessionID const): Added.
(WebDriver::WebSocketServer::session): Deleted.
* Source/WebDriver/WebSocketServer.h: Remove m_bidiService and make
  constructor private.
(WebDriver::WebSocketServer::create): Added.
* Source/WebDriver/soup/WebSocketServerSoup.cpp:
(WebDriver::WebSocketServer::disconnect): Checks and disconnect the
signal handlers.
(WebDriver::WebSocketServer::disconnectSession): Ditto.

Canonical link: https://commits.webkit.org/294135@main
